### PR TITLE
Add throwing exceptions during deserialization

### PR DESCRIPTION
**Pull request by BeksOmega**
## The basics

- [X] I branched from **project-cereal**
- [X] My pull request is against **project-cereal**
- [X] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)

Link for Diff: https://github.com/BeksOmega/blockly/compare/cereal/touch-ups...BeksOmega:cereal/exceptions
### Resolves

Work on project cereal
Dependent on #5181

### Proposed Changes

Adds throwing exceptions during deserialization as defined in the design document [here](https://docs.google.com/document/d/1Dbpezxsd4H1S0p31ncrZ1hLaMfFYBeanv8pqjkpxoHI/edit?resourcekey=0-hYVVlo828cnLmnlj30Lysg#heading=h.m045kogtm8cm).

### Reason for Changes

This makes it easier for outside devs to tell what's going wrong with their deserialization. This is especially important if they're doing any programmatic manipulation.

### Test Coverage

Added tests to ensure that errors get thrown when we expect them.

### CODE REVIEW

File: core/serialization/blocks.js [hunk 1]
```diff
@@ -383,6 +384,48 @@ const loadExtraState = function(block, state) {
   block.loadExtraState(state['extraState']);
 };
 
+/**
+ * Attempts to connect the block to the parent connection, if it exists.
+ * @param {?Connection} parentConnection The parent connnection to try to
```
Comment (BeksOmega) [inline core/serialization/blocks.js:389]: Fix Type

File: core/serialization/blocks.js [hunk 2]
```diff
@@ -18,6 +18,8 @@ const Block = goog.requireType('Blockly.Block');
 // eslint-disable-next-line no-unused-vars
 const Connection = goog.requireType('Blockly.Connection');
 const Events = goog.require('Blockly.Events');
+const {MissingBlockType, MissingConnection, BadConnectionCheck} =
+    goog.require('Blockly.serialization.exceptions');
```
Comment (BeksOmega) [inline core/serialization/blocks.js:22]: Move these into different files?

File: core/serialization/exceptions.js [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+}
+exports.MissingBlockType = MissingBlockType;
+
+/**
+ * Represents an error where deserialization encountered a block that did
+ * not have a connection that was defined in the serialized state.
+ */
+class MissingConnection extends DeserializationError {
+  /**
+   * @param {string} connection The name of the connection that is missing. Eg
```
Comment (NeilFraser) [inline core/serialization/exceptions.js:49]: E.g.

File: core/serialization/exceptions.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+class MissingBlockType extends DeserializationError {
+  /**
+   * @param {!State} state The state object which is missing the block type.
+   */
+  constructor(state) {
+    super(`Expected to find a 'type' property, defining the block type`);
+
+    /**
+     * The state object containing the bad name.
+     * @type{!State}
```
Comment (NeilFraser) [inline core/serialization/exceptions.js:36]: The rest of the codebase appears to have a space after '@type'.  Six instances in this PR.

File: core/serialization/exceptions.js [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+/**
+ * Represents an error where deserialization tried to connect two connections
+ * that were not compatible.
+ */
+class BadConnectionCheck extends DeserializationError {
+  /**
+   * @param {string} reason The reason the connections were not compatible.
+   * @param {string} childConnection The name of the incompatible child
+   *     connection. Eg 'output' or 'previous'.
```
Comment (NeilFraser) [inline core/serialization/exceptions.js:81]: E.g.

**Review by NeilFraser — APPROVED** [review]

**Pull request merged**